gc/src/lib.rs

```diff
@@ -30,17 +32,13 @@ pub use gc::force_collect;
 ///
 /// See the [module level documentation](./) for more details.
 pub struct Gc<T: Trace + ?Sized + 'static> {
-    // XXX We can probably take advantage of alignment to store this
-    root: Cell<bool>,
-    _ptr: NonZero<*mut GcBox<T>>,
+    _ptr: UnsafeCell<NonZero<*mut GcBox<T>>>,
```

Comment:
I think it should be fine.  In a case where the compiler uses it to store other data, you won't be able to flip the bit, and in a case where it has legitimate (non-zero) data, flipping the bottom bit won't have any impact on the "zeroness" because of the pointer alignment, the bottom bit wouldn't be the deciding factor on if the pointer is zeroed or not.

That said, I'm not certain what kind of optimizations the compiler does with NonZero.